apps/src/dance/ai/DanceAiModal.tsx

```diff
@@ -126,66 +186,105 @@ const DanceAiModal: React.FunctionComponent = () => {
     }
   };
 
-  const handleResultsClick = () => {
-    setMode(Mode.RESULTS);
-  };
-
-  const handleResultsFinalClick = () => {
-    setMode(Mode.RESULTS_FINAL);
-  };
-
   const handleGenerateClick = () => {
     startAi();
     setMode(Mode.GENERATING);
   };
 
-  const handleRegenerateClick = () => {
-    // Reset state
-    setTypingDone(false);
-    setResultJson('');
-    setShowPreview(false);
+  const handleStartOverClick = () => {
+    setInputs([]);
+    setCurrentInputSlot(0);
     setProcessingDone(false);
-    setGeneratingNodesDone(false);
-    setGeneratingDone(false);
+    setGeneratingProgress({step: 0, subStep: 0});
+    setMode(Mode.SELECT_INPUTS);
+  };
 
+  const handleRegenerateClick = () => {
+    setProcessingDone(false);
+    setGeneratingProgress({step: 0, subStep: 0});
     handleGenerateClick();
   };
 
   const handleProcessClick = () => {
     setMode(Mode.PROCESSING);
     setTimeout(() => {
-      setProcessingDone(true);
+      handleGenerateClick();
     }, 4000);
   };
 
+  const handleExplanationClick = () => {
+    setMode(Mode.EXPLANATION);
+  };
+
+  const handleUseClick = () => {
+    currentAiModalField?.setValue(
+      JSON.stringify({
+        inputs,
+        ...generatedEffects.current.goodEffect?.results,
+      })
+    );
+    onClose();
+  };
+
+  const handleLeaveExplanation = () => {
+    setMode(Mode.RESULTS);
+  };
+
+  // Animate through the generating process.
+  useInterval(
+    () => {
+      if (mode === Mode.GENERATING) {
+        // We do a deep copy into a new object to ensure that
+        // a re-render is triggered at the end of this work.
+        const currentGeneratingProgress: GeneratingProgress = {
+          ...generatingProgress,
+        };
```

Comment:
I don't know if this would actually be any better, but as an alternative to deep copying, you could also pass a function to the `setGeneratingProgress` state hook - something like
```
const calculateNextStep = (progress: GeneratingProgress) => {
  if (...) { 
    progress.step++;
  } 
  // etc...
  return progress;
}

setGeneratingProgress(calculateNextStep);
```

Reply:
Interestingly, a deep copy is still required with this approach too.  That said, I like splitting out into a separate function, so will do that.